Port ResourceResponse to the new CoreIPC serialization format.

gavin-apple opened the pull request:
#### d80ae68ecc7a4a5756a38de4a2eb3d2360375012
<pre>
Port ResourceResponse to the new CoreIPC serialization format.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249130">https://bugs.webkit.org/show_bug.cgi?id=249130</a>
&lt;rdar://problem/103247059&gt;

Reviewed by Alex Christensen.

Port ResourceResponse and related classes to the new serialization format.
The types moved are:
    - WebCore::ResourceRequestCachePolicy
    - WebCore::ResourceRequestBase::SameSiteDisposition
    - WebCore::ResourceRequestRequester
    - WebCore::ResourceLoadPriority
    - WebCore::ResourceResponseBase
    - WebCore::ResourceResponseBase::Type
    - WebCore::ResourceResponseBase::Tainting
    - WebCore::ResourceResponseBase::Source
    - WebCore::ReferrerPolicy

* Source/WebCore/platform/ReferrerPolicy.h:
* Source/WebCore/platform/network/ResourceLoadPriority.h:
* Source/WebCore/platform/network/ResourceRequestBase.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase):
(WebCore::m_mimeType):
(WebCore::m_textEncodingName):
(WebCore::m_httpStatusText):
(WebCore::m_httpVersion):
(WebCore::m_httpHeaderFields):
(WebCore::m_networkLoadMetrics):
(WebCore::m_httpStatusCode):
(WebCore::ResourceResponseBase::getResponseData const):
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::encode const):
(WebCore::ResourceResponseBase::decode):
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebCore/platform/network/curl/ResourceResponse.h:
* Source/WebCore/platform/network/soup/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebKit/NetworkProcess/cache/CacheStorageEngineCache.cpp:
(WebKit::CacheStorage::decodeDOMCacheRecord):
* Source/WebKit/NetworkProcess/cache/NetworkCacheEntry.cpp:
(WebKit::NetworkCache::Entry::decodeStorageRecord):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/257908@main">https://commits.webkit.org/257908@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/336e904899b65407290eda175319ef11cc7f452c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100217 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9384 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109544 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169777 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10285 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92638 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107433 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105993 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91058 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34460 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22451 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77406 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3143 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23968 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3121 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9249 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43456 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5431 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4953 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -81,11 +81,34 @@ ResourceResponseBase::ResourceResponseBase(const URL& url, const String& mimeTyp
+}
+
+ResourceResponseBase::ResourceResponseBase(std::optional<ResourceResponseBase::ResponseData> data)
+    : m_url(data.has_value() ? data->m_url : URL { })
```
.has_value() can be removed from all of these.

achristensen07 commented on Source/WebCore/platform/network/ResourceResponseBase.cpp:
```diff
@@ -81,11 +81,34 @@ ResourceResponseBase::ResourceResponseBase(const URL& url, const String& mimeTyp
+    , m_httpHeaderFields(data.has_value() ? data->m_httpHeaderFields : HTTPHeaderMap { })
+    , m_networkLoadMetrics(data.has_value() ? data->m_networkLoadMetrics : Box<WebCore::NetworkLoadMetrics> { })
+    , m_certificateInfo(data.has_value() ? data->m_certificateInfo : std::nullopt)
+    , m_haveParsedCacheControlHeader(false)
```
Can these be a an initializer in the header instead?  They doesn't depend on data.

gavin-apple committed:
Port ResourceResponse to the new CoreIPC serialization format.
https://bugs.webkit.org/show_bug.cgi?id=249130
<rdar://problem/103247059>

Reviewed by Alex Christensen.

Port ResourceResponse and related classes to the new serialization format.
The types moved are:
    - WebCore::ResourceRequestCachePolicy
    - WebCore::ResourceRequestBase::SameSiteDisposition
    - WebCore::ResourceRequestRequester
    - WebCore::ResourceLoadPriority
    - WebCore::ResourceResponseBase
    - WebCore::ResourceResponseBase::Type
    - WebCore::ResourceResponseBase::Tainting
    - WebCore::ResourceResponseBase::Source
    - WebCore::ReferrerPolicy

* Source/WebCore/platform/ReferrerPolicy.h:
* Source/WebCore/platform/network/ResourceLoadPriority.h:
* Source/WebCore/platform/network/ResourceRequestBase.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase):
(WebCore::m_mimeType):
(WebCore::m_textEncodingName):
(WebCore::m_httpStatusText):
(WebCore::m_httpVersion):
(WebCore::m_httpHeaderFields):
(WebCore::m_networkLoadMetrics):
(WebCore::m_httpStatusCode):
(WebCore::ResourceResponseBase::getResponseData const):
* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseBase::encode const):
(WebCore::ResourceResponseBase::decode):
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebCore/platform/network/curl/ResourceResponse.h:
* Source/WebCore/platform/network/soup/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebKit/NetworkProcess/cache/CacheStorageEngineCache.cpp:
(WebKit::CacheStorage::decodeDOMCacheRecord):
* Source/WebKit/NetworkProcess/cache/NetworkCacheEntry.cpp:
(WebKit::NetworkCache::Entry::decodeStorageRecord):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/257908@main
